pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,367 @@
+
+	err = createOrUpdateNetworkSecGroup(ctx, factory, opt, networkSecGroupResp)
+	if err != nil {
+		return err
+	}
+	logger.Info("create bastion network security group rules: ", "network", opt.SecurityGroupName)
+
+	return nil
+}
+
+func ensurePublicIPAddress(ctx context.Context, factory azureclient.Factory, opt *Options) (*network.PublicIPAddress, error) {
+	ip, err := getPublicIP(ctx, factory, opt)
+	if err != nil {
+		return nil, err
+	}
+	if ip != nil {
+		return ip, nil
+	}
+
+	logger.Info("create new bastion compute instance public ip address")
+	parameters := publicIPAddressDefine(opt)
+
+	ip, err = createPublicIP(ctx, factory, opt, parameters)
+	if err != nil {
+		return nil, err
+	}
+	if ip != nil {
+		return ip, nil
+	}
+
+	return nil, fmt.Errorf("failed to get (create) public ip address: %w", err)
+}
+
+func ensureComputeInstance(ctx context.Context, logger logr.Logger, bastion *extensionsv1alpha1.Bastion, factory azureclient.Factory, opt *Options) (*compute.VirtualMachine, error) {
+	instance, err := getBastionInstance(ctx, factory, opt)
+	if instance != nil || err != nil {
+		return instance, err
+	}
+
+	logger.Info("creating new bastion compute instance")
+
+	randomPassword, err := password()
+	if err != nil {
+		return nil, fmt.Errorf("%w, failed to generate random password for compute instance", err)
+	}
+
+	parameters := computeInstanceDefine(opt, bastion, randomPassword)
+
+	instance, err = createBastionInstance(ctx, factory, opt, parameters)
+	if err != nil {
+		return nil, fmt.Errorf("%w, failed to create bastion compute instance", err)
+	}
+	if instance != nil {
+		return instance, err
+	}
+
+	return nil, fmt.Errorf("%w, failed to get / create bastion compute instance", err)
+}
+
+func ensureNic(ctx context.Context, factory azureclient.Factory, opt *Options, ip *network.PublicIPAddress) (*network.Interface, error) {
+	nic, err := getNic(ctx, factory, opt)
+	if err != nil {
+		return nil, err
+	}
+	if nic != nil {
+		return nic, nil
+	}
+
+	logger.Info("create new bastion compute instance nic")
+
+	subnet, err := getSubnet(ctx, factory, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	parameters := nicDefine(opt, ip, subnet)
+
+	nicClient, err := factory.NetworkInterface(ctx, opt.SecretReference)
+	if err != nil {
+		return nil, err
+	}
+
+	if nic, err = nicClient.CreateOrUpdate(ctx, opt.ResourceGroupName, opt.NicName, *parameters); err != nil {
+		return nil, fmt.Errorf("failed to create bastion compute nic: %w", err)
+	}
+	if nic != nil {
+		return nic, nil
+	}
+
+	return nil, fmt.Errorf("failed to get (create) bastion compute nic: %w", err)
+}
+
+func getInstanceEndpoints(instance *compute.VirtualMachine, nic *network.Interface, publicIp *network.PublicIPAddress) (*bastionEndpoints, error) {
+	if instance == nil {
+		return nil, fmt.Errorf("compute instance can't be nil")
+	}
+
+	if *instance.VirtualMachineProperties.ProvisioningState != "Succeeded" {
```

Comment:
Should this things not rather be checked in `ensureComputeInstance()`?
Then we could assume here that the instance is always there.

Reply:
fixed https://github.com/gardener/gardener-extension-provider-azure/pull/397/commits/b13c4747aa3f139df3bfca3b58dd33a7aa5d1b9b

Reply:
https://github.com/gardener/gardener-extension-provider-azure/pull/397/commits/d9df6dae49635d4a3641ab6a59fa9da0820071f4

Reply:
3ccbfe1e